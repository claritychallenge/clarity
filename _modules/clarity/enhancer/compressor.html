

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>clarity.enhancer.compressor &mdash; pyClarity 0.7.1.post21 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/custom.css" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=5cc28f93"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script type="module" src="https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs"></script>
      <script type="module" src="https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs"></script>
      <script type="module">import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";import elkLayouts from "https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs";mermaid.registerLayoutLoaders(elkLayouts);mermaid.initialize({startOnLoad:false});</script>
      <script src="https://cdn.jsdelivr.net/npm/d3@7.9.0/dist/d3.min.js"></script>
      <script type="module">
import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";
window.addEventListener("load", () => mermaid.run());
</script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            pyClarity
              <img src="../../../_static/challenges.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">pyClarity</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">clarity.enhancer.compressor</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for clarity.enhancer.compressor</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Compressor Class&quot;&quot;&quot;</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">__future__</span><span class="w"> </span><span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Final</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="n">EPS</span><span class="p">:</span> <span class="n">Final</span> <span class="o">=</span> <span class="mf">1e-8</span>


<div class="viewcode-block" id="Compressor">
<a class="viewcode-back" href="../../../clarity.enhancer.html#clarity.enhancer.compressor.Compressor">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">Compressor</span><span class="p">:</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">fs</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">44100.0</span><span class="p">,</span>
        <span class="n">attack</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">5.0</span><span class="p">,</span>
        <span class="n">release</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">20.0</span><span class="p">,</span>
        <span class="n">threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
        <span class="n">attenuation</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0001</span><span class="p">,</span>
        <span class="n">rms_buffer_size</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
        <span class="n">makeup_gain</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span>
        <span class="o">**</span><span class="n">_kwargs</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Instantiate the Compressor Class.</span>

<span class="sd">        Args:</span>
<span class="sd">            fs (float): (default = 44100.0)</span>
<span class="sd">            attack (float): (default = 5.0)</span>
<span class="sd">            release float: (default = 20.0)</span>
<span class="sd">            threshold (float): (default = 1.0)</span>
<span class="sd">            attenuation (float): (default = 0.0001)</span>
<span class="sd">            rms_buffer_size (float): (default = 0.2)</span>
<span class="sd">            makeup_gain (float): (default = 1.0)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fs</span> <span class="o">=</span> <span class="n">fs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rms_buffer_size</span> <span class="o">=</span> <span class="n">rms_buffer_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_attack</span><span class="p">(</span><span class="n">attack</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_release</span><span class="p">(</span><span class="n">release</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">threshold</span> <span class="o">=</span> <span class="n">threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">attenuation</span> <span class="o">=</span> <span class="n">attenuation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">makeup_gain</span> <span class="o">=</span> <span class="n">makeup_gain</span>

        <span class="c1"># window for computing rms</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">win_len</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rms_buffer_size</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">fs</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">window</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">win_len</span><span class="p">)</span>

<div class="viewcode-block" id="Compressor.set_attack">
<a class="viewcode-back" href="../../../clarity.enhancer.html#clarity.enhancer.compressor.Compressor.set_attack">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_attack</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">t_msec</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;DESCRIPTION</span>

<span class="sd">        Args:</span>
<span class="sd">            t_msec (float): DESCRIPTION</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: DESCRIPTION</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">t_sec</span> <span class="o">=</span> <span class="n">t_msec</span> <span class="o">/</span> <span class="mf">1000.0</span>
        <span class="n">reciprocal_time</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">t_sec</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">attack</span> <span class="o">=</span> <span class="n">reciprocal_time</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fs</span></div>


<div class="viewcode-block" id="Compressor.set_release">
<a class="viewcode-back" href="../../../clarity.enhancer.html#clarity.enhancer.compressor.Compressor.set_release">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">set_release</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">t_msec</span><span class="p">:</span> <span class="nb">float</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;DESCRIPTION</span>

<span class="sd">        Args:</span>
<span class="sd">            t_msec (float): DESCRIPTION</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: DESCRIPTION</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">t_sec</span> <span class="o">=</span> <span class="n">t_msec</span> <span class="o">/</span> <span class="mf">1000.0</span>
        <span class="n">reciprocal_time</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">t_sec</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">release</span> <span class="o">=</span> <span class="n">reciprocal_time</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">fs</span></div>


<div class="viewcode-block" id="Compressor.process">
<a class="viewcode-back" href="../../../clarity.enhancer.html#clarity.enhancer.compressor.Compressor.process">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">signal</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="nb">list</span><span class="p">[</span><span class="n">Any</span><span class="p">]]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;DESCRIPTION</span>

<span class="sd">        Args:</span>
<span class="sd">            signal (np.array): DESCRIPTION</span>

<span class="sd">        Returns:</span>
<span class="sd">            np.array: DESCRIPTION</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">padded_signal</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">win_len</span> <span class="o">-</span> <span class="mi">1</span><span class="p">),</span> <span class="n">signal</span><span class="p">))</span>
        <span class="n">rms</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">convolve</span><span class="p">(</span><span class="n">padded_signal</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">window</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;valid&quot;</span><span class="p">)</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">win_len</span>
            <span class="o">+</span> <span class="n">EPS</span>
        <span class="p">)</span>
        <span class="n">comp_ratios</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">curr_comp</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">for</span> <span class="n">rms_i</span> <span class="ow">in</span> <span class="n">rms</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">rms_i</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">threshold</span><span class="p">:</span>
                <span class="n">temp_comp</span> <span class="o">=</span> <span class="p">(</span><span class="n">rms_i</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">attenuation</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span>
                    <span class="p">(</span><span class="mf">1.0</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">attenuation</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">threshold</span>
                <span class="p">)</span>
                <span class="n">curr_comp</span> <span class="o">=</span> <span class="n">curr_comp</span> <span class="o">*</span> <span class="p">(</span><span class="mf">1.0</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">attack</span><span class="p">)</span> <span class="o">+</span> <span class="p">(</span><span class="n">temp_comp</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">attack</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">curr_comp</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">release</span> <span class="o">+</span> <span class="n">curr_comp</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">release</span><span class="p">)</span>
            <span class="n">comp_ratios</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_comp</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">signal</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">comp_ratios</span><span class="p">)</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">makeup_gain</span><span class="p">),</span> <span class="n">rms</span><span class="p">,</span> <span class="n">comp_ratios</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2025, pyClarity authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>