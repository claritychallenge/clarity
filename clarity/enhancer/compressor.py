"""Compressor Class"""
import numpy as np


class Compressor:
    def __init__(
        self,
        fs: int = 44100,
        attack: int = 5,
        release: int = 20,
        threshold: int = 1,
        attenuation: float = 0.0001,
        rms_buffer_size: float = 0.2,
        makeup_gain: int = 1,
    ):
        """Instantiate the Compressor Class.

        Args:
            fs (int): (default = 44100)
            attack (int): (default = 5)
            release int: (default = 20)
            threshold (int): (default = 1)
            attenuation (float): (default = 0.0001)
            rms_buffer_size (float): (default = 0.2)
            makeup_gain (int): (default = 1)
        """
        self.fs = fs
        self.rms_buffer_size = rms_buffer_size
        self.set_attack(attack)
        self.set_release(release)
        self.threshold = threshold
        self.attenuation = attenuation
        self.eps = 1e-8
        self.makeup_gain = makeup_gain

        # window for computing rms
        self.win_len = int(self.rms_buffer_size * self.fs)
        self.window = np.ones(self.win_len)

    def set_attack(self, t_msec: float) -> float:
        """DESCRIPTION

        Args:
            t_msec (float): DESCRIPTION

        Returns:
            float: DESCRIPTION
        """
        t_sec = t_msec / 1000
        reciprocal_time = 1 / t_sec
        self.attack = reciprocal_time / self.fs

    def set_release(self, t_msec: float) -> float:
        """DESCRIPTION

        Args:
            t_msec (float): DESCRIPTION

        Returns:
            float: DESCRIPTION
        """
        t_sec = t_msec / 1000
        reciprocal_time = 1 / t_sec
        self.release = reciprocal_time / self.fs

    def process(self, signal: np.array) -> np.array:
        """DESCRIPTION

        Args:
            signal (np.array): DESCRIPTION

        Returns:
            np.array: DESCRIPTION
        """
        padded_signal = np.concatenate((np.zeros(self.win_len - 1), signal))
        rms = np.sqrt(
            np.convolve(padded_signal**2, self.window, mode="valid") / self.win_len
            + self.eps
        )
        comp_ratios = []
        curr_comp = 1
        for rms_i in rms:
            if rms_i > self.threshold:
                temp_comp = (rms_i * self.attenuation) + (
                    (1 - self.attenuation) * self.threshold
                )
                curr_comp = (curr_comp * (1 - self.attack)) + (temp_comp * self.attack)
            else:
                curr_comp = (1 * self.release) + curr_comp * (1 - self.release)
            comp_ratios.append(curr_comp)
        return signal * np.array(comp_ratios) * self.makeup_gain, rms, comp_ratios
