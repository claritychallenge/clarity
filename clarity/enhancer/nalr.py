import math
from typing import Optional, Tuple, Union

import numpy as np
import scipy
import scipy.signal


def firwin2(
    n: int,
    f: np.ndarray,
    a: np.ndarray,
    window: Optional[Union[str, float]] = None,
    antisymmetric: Optional[bool] = None,
) -> np.ndarray:
    """FIR filter design using the window method.
    Partial implementation of scipy firwin2 but using our own MATLAB-derived fir2.

    Args:
        n (int): The number of taps in the FIR filter.
        f (ndarray): The frequency sampling points. 0.0 to 1.0 with 1.0 being Nyquist.
        a (ndarray): The filter gains at the frequency sampling points.
        window (string or (string, float), optional): See scipy.firwin2 (default: (None))
        _antisymmetric (bool, optional): Unused but present to main compatability
            with scipy firwin2.
    Returns:
        np.ndarray:  The filter coefficients of the FIR filter, as a 1-D array of length n.
    """
    window_shape = None
    window_type: Union[str, float, None]
    window_param: Union[str, float]
    if isinstance(window, tuple):
        window_type, window_param = window if window is not None else (None, 0)
    else:
        window_type = window

    order = n - 1

    if window_type == "kaiser":
        window_shape = scipy.signal.kaiser(n, window_param)

    if window_shape is None:
        b, _ = fir2(order, f, a)
    else:
        b, _ = fir2(order, f, a, window_shape)

    return b


def fir2(
    nn: int, ff: np.ndarray, aa: np.ndarray, npt: Optional[int] = None
) -> Tuple[np.ndarray, int]:
    """FIR arbitrary shape filter design using the frequency sampling method.
    Translation of MATLAB fir2.

    Args:
        nn (int): Order
        ff (np.ndarray): Frequency breakpoints (0 < F < 1) where 1 is Nyquist rate.
                        First and last elements must be 0 and 1 respectively
        aa (np.ndarray): Magnitude breakpoints
        npt (int, optional): Number of points for freq response interpolation
            (default: max (smallest power of 2 greater than nn, 512))

    Returns:
        np.ndarray: nn + 1 filter coefficients
    """
    # Work with filter length instead of filter order
    nn += 1

    if npt is None:
        npt = 2.0 ** np.ceil(math.log(nn) / math.log(2)) if nn >= 1024 else 512
        wind = scipy.signal.hamming(nn)
    else:
        wind = npt
        npt = 2.0 ** np.ceil(math.log(nn) / math.log(2)) if nn >= 1024 else 512
    lap = int(np.fix(npt / 25))

    nbrk = max(len(ff), len(aa))

    ff[0] = 0
    ff[nbrk - 1] = 1

    H = np.zeros(npt + 1)
    nint: int = nbrk - 1
    df = np.diff(ff, n=1)

    npt += 1
    nb: int = 0
    H[0] = aa[0]

    for i in np.arange(nint):
        if df[i] == 0:
            nb = int(np.ceil(nb - lap / 2))
            ne = nb + lap - 1
        else:
            ne = int(np.fix(ff[i + 1] * npt)) - 1

        j = np.arange(nb, ne + 1)
        inc: Union[float, np.ndarray] = 0.0 if nb == ne else (j - nb) / (ne - nb)
        H[nb : (ne + 1)] = (inc * aa[i + 1]) + ((1 - inc) * aa[i])
        nb = ne + 1

    dt: float = 0.5 * (nn - 1)
    rad = -dt * 1j * math.pi * np.arange(0, npt) / (npt - 1)
    H = H * np.exp(rad)

    H = np.concatenate((H, H[npt - 2 : 0 : -1].conj()))
    ht = np.real(np.fft.ifft(H))

    b = ht[0:nn] * wind

    return b, 1


class NALR:
    def __init__(self, nfir: int, fs: int):
        """
        Args:
            nfir: Order of the NAL-R EQ filter and the matching delay
            fs: Sampling rate in Hz
        """
        self.nfir = nfir
        # Processing parameters
        self.fmax = 0.5 * fs

        # Audiometric frequencies
        self.aud = np.array([250, 500, 1000, 2000, 4000, 6000])

        # Design a flat filter having the same delay as the NAL-R filter
        self.delay = np.zeros(nfir + 1)
        self.delay[nfir // 2] = 1.0

    def hl_interp(self, hl: np.ndarray, cfs: np.ndarray):
        try:
            hl_interpf = scipy.interpolate.interp1d(cfs, hl)
        except ValueError:
            raise ValueError(
                "Hearing losses (hl) and center frequencies (cfs) don't match!"
            )
        return hl_interpf(self.aud)

    def build(
        self,
        hl: np.ndarray,
        cfs: np.ndarray = None,
    ):
        """
        Args:
            hl: hearing thresholds at [250, 500, 1000, 2000, 4000, 6000] Hz
            cfs: center frequencies of the hearing thresholds. If None, the default
                values are used.
        Returns:
            NAL-R FIR filter
            delay
        """
        if cfs is None:
            cfs = np.array([250, 500, 1000, 2000, 3000, 6000])

        hl = self.hl_interp(np.array(hl), np.array(cfs))
        mloss = np.max(hl)

        if mloss > 0:
            # Compute the NAL-R frequency response at the audiometric frequencies
            bias = np.array([-17, -8, 1, -1, -2, -2])
            t3 = hl[1] + hl[2] + hl[3]
            if t3 <= 180:
                xave = 0.05 * t3
            else:
                xave = 9.0 + 0.116 * (t3 - 180)
            gdB = xave + 0.31 * hl + bias
            gdB = np.clip(gdB, a_min=0, a_max=None)

            # Design the linear-phase FIR filter
            fv = np.append(
                np.append(0, self.aud), self.fmax
            )  # Frequency vector for the interpolation
            cfreq = (
                np.linspace(0, self.nfir, self.nfir + 1) / self.nfir
            )  # Uniform frequency spacing from 0 to 1
            gdBv = np.append(
                np.append(gdB[0], gdB), gdB[-1]
            )  # gdB vector for the interpolation
            interpf = scipy.interpolate.interp1d(fv, gdBv)
            gain = interpf(self.fmax * cfreq)
            glin = np.power(10, gain / 20.0)
            nalr = firwin2(self.nfir + 1, cfreq, glin)
        else:
            nalr = self.delay.copy()
        return nalr, self.delay

    def apply(self, nalr: np.ndarray, wav: np.ndarray):
        """
        Args:
            nalr: built NAL-R FIR filter
            wav: one dimensional wav signal

        Returns:
            amplified signal
        """
        return np.convolve(wav, nalr)
