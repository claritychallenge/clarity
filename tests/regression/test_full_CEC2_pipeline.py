"""Regression Tests for CEC2"""
# pylint: disable=too-many-locals invalid-name

# Pass some random data through code and compare with reference output
# scene_renderer, enhancer, compressor, haspi
from typing import Optional

import numpy as np
from cpuinfo import get_cpu_info
from omegaconf import OmegaConf
from scipy.io import wavfile

from clarity.data.scene_renderer_cec2 import SceneRenderer
from clarity.enhancer.compressor import Compressor
from clarity.enhancer.nalr import NALR
from clarity.evaluator.haspi import haspi_v2_be

CPUINFO = get_cpu_info()

np.random.seed(0)

# Set up some scene to simulate
# It's been designed to get good code coverage but running quickly
# - Using three maskers - one from each noise type
# - Using a short target with reduce pre and post silence
# - Only generating 2 hearing aid channels
SCENE = {
    "dataset": "demo",
    "room": "R06001",
    "scene": "S06001",
    "target": {"name": "T010_G0N_02468", "time_start": 0, "time_end": 115894},
    "duration": 8820,
    "interferers": [
        {
            "position": 1,
            "time_start": 0,
            "time_end": 8820,
            "type": "noise",
            "name": "CIN_fan_014.wav",
            "offset": 5376,
        },
        {
            "position": 2,
            "time_start": 0,
            "time_end": 8820,
            "type": "speech",
            "name": "som_04766_05.wav",
            "offset": 40000,
        },
        {
            "position": 3,
            "time_start": 0,
            "time_end": 8820,
            "type": "music",
            "name": "1111967.low.mp3",
            "offset": 842553,
        },
    ],
    "SNR": 0.0,
    "listener": {
        "rotation": [
            {"sample": 100, "angle": 52.3628},
            {"sample": 400, "angle": 38.5256},
        ],
        "hrir_filename": ["VP_N6-ED", "VP_N6-BTE_fr"],
    },
}

DEMO_PATHS = OmegaConf.create(
    {
        "hoairs": "tests/test_data/rooms/HOA_IRs",
        "hrirs": "tests/test_data/hrir/HRIRs_MAT",
        "scenes": "tests/test_data/clarity_data/demo/scenes",
        "targets": "tests/test_data/targets",
        "interferers": "tests/test_data/interferers/{type}",
    }
)

DEMO_METADATA = OmegaConf.create(
    {
        "room_definitions": "tests/test_data/metadata/rooms.demo.json",
        "scene_definitions": "",  # Scene definition file not needed for test
        "hrir_metadata": "tests/test_data/metadata/hrir_data.json",
    }
)

SCENE_RENDERER = SceneRenderer(
    DEMO_PATHS,
    DEMO_METADATA,
    ambisonic_order=6,
    equalise_loudness=True,
    reference_channel=1,
    channel_norms=[12.0, 3.0],
)


def test_full_cec2_pipeline(
    regtest,
    tmp_path,
    scene: Optional[dict] = None,
    _demo_paths: OmegaConf = DEMO_PATHS,
    _demo_metadata: OmegaConf = DEMO_METADATA,
    scene_renderer: SceneRenderer = SCENE_RENDERER,
) -> None:
    """Test full CEC2 pipeline"""

    if scene is None:
        scene = SCENE
    target, interferers, anechoic, _head_turn = scene_renderer.generate_hoa_signals(
        scene
    )

    scene_renderer.generate_binaural_signals(
        scene, target, interferers, anechoic, str(tmp_path)
    )
    _, reference = wavfile.read(f"{tmp_path}/S06001_target_anechoic_CH1.wav")
    _, signal = wavfile.read(f"{tmp_path}/S06001_mix_CH1.wav")

    reference = reference.astype(float) / 32768.0
    signal = signal.astype(float) / 32768.0

    # Truncate to 200 ms - i.e. just use part of the signals
    # to speed up the HASPI calculation a little
    signal = signal[:8820, :]
    reference = reference[:8820, :]

    # The data below doesn't really need to be meaningful.
    # The purpose of the test is not to see if the haspi score is reasonable
    # but just to check that the results do not change unexpectedly across releases.

    nalr_cfg = {"nfir": 220, "fs": 44100}
    compressor_cfg = {
        "threshold": 0.35,
        "attenuation": 0.1,
        "attack": 50,
        "release": 1000,
        "rms_buffer_size": 0.064,
    }

    audiogram_l = [45, 50, 60, 65, 60, 65, 70, 80]
    audiogram_r = [45, 45, 60, 70, 60, 60, 80, 80]
    audiogram_cfs = [250, 500, 1000, 2000, 3000, 4000, 6000, 8000]

    fs = 44100

    enhancer = NALR(**nalr_cfg)
    compressor = Compressor(**compressor_cfg)  # type: ignore

    nalr_fir, _ = enhancer.build(audiogram_l, audiogram_cfs)
    out_l = enhancer.apply(nalr_fir, signal[:, 0])

    nalr_fir, _ = enhancer.build(audiogram_r, audiogram_cfs)
    out_r = enhancer.apply(nalr_fir, signal[:, 1])

    out_l, _, _ = compressor.process(out_l)
    out_r, _, _ = compressor.process(out_r)

    enhanced_audio = np.stack([out_l, out_r], axis=1)

    enhanced_audio = np.tanh(enhanced_audio)

    sii_enhanced = haspi_v2_be(
        reference_left=reference[:, 0],
        reference_right=reference[:, 1],
        processed_left=enhanced_audio[:, 0],
        processed_right=enhanced_audio[:, 1],
        fs_signal=fs,
        audiogram_left=audiogram_l,
        audiogram_right=audiogram_r,
        audiogram_cfs=audiogram_cfs,
    )

    regtest.write(f"Enhanced audio HASPI score is {sii_enhanced:0.7f}\n")

    # Enhanced audio HASPI score is 0.2994066
