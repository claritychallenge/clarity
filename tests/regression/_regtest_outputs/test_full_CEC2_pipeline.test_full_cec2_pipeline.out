Enhanced audio HASPI score is 0.7491529
